payment/src/main/java/org/killbill/billing/payment/core/PaymentProcessor.java

```diff
@@ -472,53 +483,53 @@ public boolean apply(final PaymentTransactionModelDao curPaymentTransactionModel
-    }
+    private List<PaymentAttempt> getPaymentAttempts(final List<PaymentTransaction> purchasedTransactions,
+                                                    final List<PaymentAttemptModelDao> paymentAttempts,
+                                                    final InternalTenantContext internalTenantContext) {
+
+        List<PaymentAttempt> result = new ArrayList<PaymentAttempt>();
 
-    private int getNumberAttemptsInState(final Collection<PaymentTransaction> allTransactions, final TransactionStatus... statuses) {
-        if (allTransactions == null || allTransactions.size() == 0) {
-            return 0;
+        // Add Past Payment Attempts
+        for (PaymentAttemptModelDao pastPaymentAttempt : paymentAttempts) {
+            DefaultPaymentAttempt paymentAttempt = new DefaultPaymentAttempt(pastPaymentAttempt);
+            result.add(paymentAttempt);
         }
-        return Collections2.filter(allTransactions, new Predicate<PaymentTransaction>() {
-            @Override
-            public boolean apply(final PaymentTransaction input) {
-                for (final TransactionStatus cur : statuses) {
-                    if (input.getTransactionStatus() == cur) {
-                        return true;
-                    }
-                }
-                return false;
+
+        // Get Future Payment Attempts from Notification Queue and add them to the list
+        try {
+            final NotificationQueue retryQueue = notificationQueueService.getNotificationQueue(DefaultPaymentService.SERVICE_NAME, "retry");
+            final List<NotificationEventWithMetadata<NotificationEvent>> notificationEventWithMetadatas =
+                    retryQueue.getFutureNotificationForSearchKeys(internalTenantContext.getAccountRecordId(), internalTenantContext.getTenantRecordId());
+
+            for (NotificationEventWithMetadata<NotificationEvent> notificationEvent : notificationEventWithMetadatas) {
+                DefaultPaymentAttempt futurePaymentAttempt = new DefaultPaymentAttempt(
+                        null, //accountId,
+                        null, //paymentMethodId,
+                        ((PaymentRetryNotificationKey) notificationEvent.getEvent()).getAttemptId(), //id,
+                        notificationEvent.getEffectiveDate(), //createdDate,
+                        notificationEvent.getEffectiveDate(), //updatedDate,
+                        null, //paymentExternalKey,
+                        null, //transactionId,
+                        null, //transactionExternalKey,
+                        null, //transactionType,
+                        "SCHEDULED", //stateName,
+                        null, //amount,
+                        null, //currency,
+                        ((PaymentRetryNotificationKey) notificationEvent.getEvent()).getPaymentControlPluginNames().get(0), //pluginName,
+                        null);//pluginProperties
+                result.add(futurePaymentAttempt);
             }
```

Comment:
Yes, you should complete based on previous attempt. Some fields will be left null (as they are not yet known)
* `paymentMethodId`
* `transactionId`
* `createdDate`/`updatedDate` (but we should set `effectiveDate`)